vochain/account.go

```diff
@@ -461,250 +473,253 @@ func (v *State) SetAccount(accountAddress common.Address, account *Account) erro
-			sigAddress.String(),
+			txSenderAddress.String(),
 		)
 	}
-	// check to
 	txToAddress := common.BytesToAddress(tx.To)
 	toTxAccount, err := state.GetAccount(txToAddress, false)
 	if err != nil {
-		return nil, fmt.Errorf("cannot get to account info: %w", err)
+		return fmt.Errorf("cannot get to account: %w", err)
 	}
 	if toTxAccount == nil {
-		return nil, ErrAccountNotExist
+		return ErrAccountNotExist
 	}
-	// check nonce
-	acc, err := state.GetAccount(sigAddress, false)
+	acc, err := state.GetAccount(txSenderAddress, false)
 	if err != nil {
-		return nil, fmt.Errorf("cannot get account info: %w", err)
+		return fmt.Errorf("cannot get from account: %w", err)
 	}
 	if acc == nil {
-		return nil, ErrAccountNotExist
+		return ErrAccountNotExist
 	}
 	if tx.Nonce != acc.Nonce {
-		return nil, fmt.Errorf("invalid nonce, expected %d got %d", acc.Nonce, tx.Nonce)
+		return fmt.Errorf("invalid nonce, expected %d got %d", acc.Nonce, tx.Nonce)
 	}
-	// get tx cost
 	cost, err := state.TxCost(models.TxType_SEND_TOKENS, false)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	// check value
 	if (tx.Value + cost) > acc.Balance {
-		return nil, ErrNotEnoughBalance
+		return ErrNotEnoughBalance
 	}
-	return &sendTokensTxCheckValues{sigAddress, txToAddress, tx.Value, tx.Nonce}, nil
+	return nil
 }
 
 // SetAccountDelegateTxCheck checks if a SetAccountDelegateTx and its data are valid
-func SetAccountDelegateTxCheck(vtx *models.Tx, txBytes, signature []byte, state *State) (*setAccountDelegateTxCheckValues, error) {
-	if vtx == nil {
-		return nil, fmt.Errorf("transaction is nil")
+func SetAccountDelegateTxCheck(vtx *models.Tx, txBytes, signature []byte, state *State) error {
+	if vtx == nil || signature == nil || txBytes == nil || state == nil {
+		return fmt.Errorf("invalid parameters")
+	}
+	tx := vtx.GetSetAccount()
+	if tx == nil {
+		return fmt.Errorf("invalid tx")
+	}
+	if tx.Txtype != models.TxType_ADD_DELEGATE_FOR_ACCOUNT &&
+		tx.Txtype != models.TxType_DEL_DELEGATE_FOR_ACCOUNT {
+		return fmt.Errorf("invalid tx type")
 	}
-	tx := vtx.GetSetAccountDelegateTx()
-	// check signature available
-	if signature == nil || tx == nil || txBytes == nil {
-		return nil, fmt.Errorf("missing signature and/or transaction")
+	if tx.Nonce == nil {
+		return fmt.Errorf("invalid nonce")
 	}
-	// check delegate
-	delAcc := common.BytesToAddress(tx.Delegate)
-	if delAcc == types.EthereumZeroAddress {
-		return nil, fmt.Errorf("invalid delegate address")
+	if len(tx.Delegates) == 0 {
+		return fmt.Errorf("invalid delegates")
 	}
-	// get sender
-	sigAddress, acc, err := state.AccountFromSignature(txBytes, signature)
+	txSenderAddress, txSenderAccount, err := state.AccountFromSignature(txBytes, signature)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	// check delegate to add is not itself
-	if delAcc == *sigAddress {
-		return nil, fmt.Errorf("cannot add self to delegates list")
+	delegatesToSetMap := make(map[common.Address]bool)
```

Comment:
I saw this code snipped already. Is it necessary to duplicate it? In theory the TxCheck functions must be called before modifying the sate, so one can relay on the other.

Reply:
they have different purposes, one checks duplicates against tx.delegates and the other against the account that will be eventually modified.